src/scales/scale.linear.js

```diff
@@ -11,116 +11,131 @@ var defaultConfig = {
 	}
 };
 
+var DEFAULT_MIN = 0;
+var DEFAULT_MAX = 1;
+
+function getMatchingVisibleMetas(scale) {
+	var isHorizontal = scale.isHorizontal();
+
+	return scale.chart._getSortedVisibleDatasetMetas()
+		.filter(function(meta) {
+			return isHorizontal ? meta.xAxisID === scale.id : meta.yAxisID === scale.id;
+		});
+}
+
+function getOrCreateStack(stacks, stacked, meta) {
+	var key = [
+		meta.type,
+		// we have a separate stack for stack=undefined datasets when the opts.stacked is undefined
+		stacked === undefined && meta.stack === undefined ? meta.index : '',
+		meta.stack
+	].join('.');
+
+	if (stacks[key] === undefined) {
+		stacks[key] = {
+			pos: [],
+			neg: []
+		};
+	}
+
+	return stacks[key];
+}
+
+function isInvalidOrHidden(value, meta, i) {
+	return isNaN(value.min) || isNaN(value.max) || meta.data[i].hidden;
+}
+
+function updateStackAtIndex(stack, i, value, opts) {
+	var pos = stack.pos;
+	var neg = stack.neg;
+
+	pos[i] = pos[i] || 0;
+	neg[i] = neg[i] || 0;
+
+	if (opts.relativePoints) {
+		pos[i] = 100;
+	} else if (value.min < 0 || value.max < 0) {
+		neg[i] += value.min;
+	} else {
+		pos[i] += value.max;
+	}
+}
+
+function stackData(scale, stacks, meta, data) {
+	var opts = scale.options;
+	var stacked = opts.stacked;
+	var stack = getOrCreateStack(stacks, stacked, meta);
+	var ilen = data.length;
+	var i, value;
+
+	for (i = 0; i < ilen; ++i) {
+		value = scale._parseValue(data[i]);
+		if (isInvalidOrHidden(value, meta, i)) {
+			continue;
+		}
+
+		updateStackAtIndex(stack, i, value, opts);
```

Comment:
I might inline this function. It'd be more performant than making a function call for every data point and you can set `pos` and `neg` outside the loop instead of resetting for each call. I also think it makes it a little easier to follow